NebulaWorld/Factory/FactoryManager.cs

```diff
@@ -34,6 +35,30 @@ public static void Initialize()
             TargetPlanet = PLANET_NONE;
         }
 
+        public static void InitializePrebuildRequests()
+        {
+            //Load existing prebuilds to the dictionary so it will be ready to build
+            if (LocalPlayer.IsMasterClient)
+            {
+                using (GetPrebuildRequests(out var prebuildRequests))
+                {
+                    foreach (PlanetFactory factory in GameMain.data.factories)
+                    { 
+                        if (factory != null)
+                        {
+                            foreach (PrebuildData prebuild in factory.prebuildPool)
+                            {
+                                if (prebuild.id != 0)
```

Comment:
Isn't the array tail full with 0's? We could eventually break before iterating everything.

Reply:
That is true, but I am not sure if it is always in order. For example, the host will build 3 prebuilds and drones will finish the middle one. In that case, there will be:
`[-1, 0, -3, 0, 0, 0, 0]`

Reply:
Doesn't the game compact the array gaps or at least have a counter for the last prebuild ID used?

Reply:
Yup, the game uses `[0, prebuildCursor)` range for values and `prebuild[i].id == i` to check for an instance being valid.

Reply:
I tried this:
```cs
using (GetPrebuildRequests(out var prebuildRequests))
{
    foreach (PlanetFactory factory in GameMain.data.factories)
    { 
        if (factory != null)
        {
            foreach (PrebuildData prebuild in factory.prebuildPool)
            {
                if (prebuild.id == 0)
                {
                    break;
                }
                prebuildRequests[new PrebuildOwnerKey(factory.planetId, prebuild.id)] = LocalPlayer.PlayerId;
            }
        }
    }
}
```
But it did not worked. :( Let me print the content of the array...

Reply:
So, this:
```cs
string s = "";
foreach (PrebuildData prebuild in factory.prebuildPool)
{
    s += prebuild.id + ",";
    if (prebuild.id != 0)
    {
        prebuildRequests[new PrebuildOwnerKey(factory.planetId, prebuild.id)] = LocalPlayer.PlayerId;
    }
}
Debug.Log(s);
```

will produce:
![image](https://user-images.githubusercontent.com/79051050/116903019-2729d400-ac3c-11eb-9bf6-35307d49f22b.png)

Reply:
You need to iterate from 0 to prebuildcursor. The list itself does not use prebuildCursor.

Reply:
The iteration should be like

```csharp
for (int i = 0; i < prebuildCursor; i++)
{
    if (prebuildList[i] == i) { /* stuff */ }
}
```

Reply:
preBuildCursor defines the end of the list, not list.Length. might seem like micro-optimizing, but they did it for a reason.

Reply:
fixed :)